analyses/cell-type-wilms-tumor-14/scripts/utils/01_anchor_transfer_seurat_functions.R

```diff
@@ -0,0 +1,131 @@
+                            ref_obj, sample, 
+                            level = "celltype", # celltype or compartment
+                            unknown_cutoff = 0.5, ndims = 20) {
+  # perspective output files
+  filename <- file.path(results_out_dir, paste0(sample, "_", level,".pdf"))
+  filename_csv <- file.path(results_out_dir, paste0(sample, "_", level,".csv"))
+
+  
+  sample_obj <- SeuratObject::LoadSeuratRds( file.path(path_anal,"scratch","00_preprocessing_rds",paste0(sample,".rdsSeurat")) )
+
+  
+  # set row names as gene symbol, since reference obj uses gene symbol by default
+  # length(intersect(rownames(ref_obj), rownames(sample_obj))) # make sure gene symbol consistency
+  # length(intersect(VariableFeatures(ref_obj), rownames(sample_obj))) # candidate genes used for transfer
+  # unique <- sample_obj[["RNA"]]@meta.data$gene_ids[!duplicated(sample_obj[["RNA"]]@meta.data$gene_symbol)]
+  # sample_obj <- sample_obj[unique,]
+  # rownames(sample_obj) <- sample_obj[["RNA"]]@meta.data$gene_symbol
+  
+  # find anchors
+  anchors <- FindTransferAnchors(reference = ref_obj, query = sample_obj, dims = 1:ndims)
+  nanchors <- nrow(anchors@anchors)
+  # clean annotation, too few stroma, immune and endo
+  # ref_obj@meta.data <- ref_obj@meta.data %>%
+  #   mutate(annot = case_when(compartment == "stroma" ~ "stroma",
+  #                            compartment == "immune" ~ "immune",
+  #                            compartment == "endothelium" ~ "endothelium",
+  #                            TRUE ~ celltype))
+  if (level == "compartment") {
+    ref_obj@meta.data <- ref_obj@meta.data %>%
+      mutate(annot = compartment)
+  } else {
+    ref_obj@meta.data <- ref_obj@meta.data %>%
+      mutate(annot = celltype)
+  }
+  ref_obj@meta.data$annot <- factor(ref_obj@meta.data$annot)
+  # transfer labels
+  predictions <- TransferData(anchorset = anchors, refdata = ref_obj$annot, dims = 1:ndims)
+  predictions <- mutate(predictions, predicted.id = case_when(prediction.score.max < unknown_cutoff ~ "Unknown",
+                                                              TRUE ~ predicted.id))
+  sample_obj <- AddMetaData(object = sample_obj, metadata = predictions)
+  
+  
+  # anchor transfer plot
+  color <- Polychrome::glasbey.colors( length( unique(predictions$predicted.id) ) +1  )
+  color <- color[-1]
+  names(color) <- unique(predictions$predicted.id)
+  color['Unknown'] <- "gray90"
+  
+  p1 <- Seurat::DimPlot(sample_obj, reduction = "umap", group.by = "predicted.id", label = F, cols = color) +
+    ggtitle(paste0(sample))
+  p2 <- Seurat::DimPlot(sample_obj, reduction = "umap", label = T)
+  df <- sample_obj@meta.data %>%
+    dplyr::group_by(seurat_clusters, predicted.id) %>%
+    dplyr::count(name = "sum")
+  p3 <- ggplot(df, aes(x = seurat_clusters, y = sum, fill =  predicted.id)) +
+    geom_bar(width = 0.5, stat = "identity", position = "fill") +
+    scale_fill_manual(values = color)
```

Comment:
I think you can do this more simply, as `ggplot::geom_bar` will already calculate this grouped count by default
```suggestion
  p3 <- ggplot(sample_obj@meta.data, aes(x = seurat_clusters, fill =  predicted.id)) +
    geom_bar(width = 0.5, position = "fill") +
    scale_fill_manual(values = color) +
    labs(y = "Proportion of cells")
```